docs/api_overview.md

```diff
@@ -0,0 +1,471 @@
+  full duration that the plugin runs in its Wasm VM, while each stream `Context`
+  object lives for the duration of the HTTP/gRPC request with which it is
+  associated.
+  
+## Registering Contexts
+
+`RegisterContextFactory` is the mechanism for bootstrapping plugin code. Plugin
+code instantiates a `RegisterContextFactory` in a static
+variable. `RegisterContextFactory` takes as constructor params `std::function`s
+for creating new root context and stream context instances. The plugin can use
+these to create and return instances of its own subclasses of `RootContext` and
+`Context`. For example:
+
+```c++
+static RegisterContextFactory register_ExampleContext(
+    CONTEXT_FACTORY(ExampleContext), ROOT_FACTORY(ExampleRootContext),
+    "my_root_id");
+```
+
+`ROOT_FACTORY` and `CONTEXT_FACTORY` are convenience macros for lambdas that
+create instances of the specified `RootContext` and `Context` subclasses.
+        
+## `RootContext`
+
+`RootContext` instances are associated with the plugin as a whole. Accordingly,
+they receive callbacks corresponding to different events in the lifecycle of a
+plugin:
+
+* `validateConfiguration`: May be called to validate the configuration data that
+  will be passed to a plugin on startup, e.g. in the control plane.
+* `onCreate`: called when context is created.
+* `onStart`: called on plugin start.
+* `onConfigure`: called on plugin start, and any time configuration subsequently
+  changes.
+* `onDone`: called when the plugin is being shut down.
+* `onDelete`: called after the plugin has acknowledged shutdown is complete, as
+  indication to release resources.
+
+`RootContexts` can also receive other callbacks for events that are not strictly
+tied to an incoming stream:
+
+* `onTick`: called when a timer fires. See [Timers].
+* `onQueueReady`: called when data arrives on a SharedQueue. See [Shared
+  queues].
+
+`RootContext` defines methods for initiating outbound HTTP and gRPC callouts:
+
+* `httpCall`: initiates an HTTP callout (i.e. outbound HTTP request). See [HTTP
+  callouts].
+* `grpcSimpleCall`: initiates a gRPC callout (i.e. outbound gRPC call). See
+  [gRPC callouts].
+* `grpcCallHandler`: initiates a gRPC callout (i.e. outbound gRPC call). See
+  [gRPC callouts].
+* `grpcStreamHandler`: initiates a streaming gRPC callout (i.e. outbound gRPC
+  call). See [gRPC callouts].
+
+For API details, see doc comments for the `RootContext` class in
+[proxy_wasm_api.h].
+
+## `Context`
+
+`Context` instances are associated with individual streams--i.e. incoming HTTP
+or gRPC requests, or TCP streams if the network proxy is acting as an L4
+proxy. `Context` instances receive callbacks corresponding to different events
+in the lifecycle of a stream:
+
+* `onCreate`: called when handling of a new stream starts.
+* `onDone`: called when the host is done processing the stream.
+* `onLog`: called after the host is done processing the stream, if the plugin is
+  being used for access logging.
```

Comment:
I'm not sure what this means. How does a plugin sign up for access logging? Clarify that this is not related to logging described below?

Reply:
This is when configured as an Envoy AccessLogger (for example). Added a link to the relevant Envoy documentation.